src/Npgsql/Replication/Logical/TestDecoding/NpgsqlLogicalReplicationConnectionExtensions.cs

```diff
@@ -0,0 +1,118 @@
+        /// </summary>
+        /// <param name="connection">The <see cref="NpgsqlLogicalReplicationConnection"/> to use for starting replication</param>
+        /// <param name="slot">The replication slot that will be updated as replication progresses so that the server
+        /// knows which WAL segments are still needed by the standby.
+        /// </param>
+        /// <param name="walLocation">The WAL location to begin streaming at.</param>
+        /// <param name="options">The collection of options passed to the slot's logical decoding plugin.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A <see cref="Task{T}"/> representing an <see cref="IAsyncEnumerable{T}"/> that
+        /// can be used to stream WAL entries in form of <see cref="NpgsqlTestDecodingData"/> instances.</returns>
+        [PublicAPI]
+        public static Task<IAsyncEnumerable<NpgsqlTestDecodingData>> StartReplication(
+            this NpgsqlLogicalReplicationConnection connection, NpgsqlTestDecodingReplicationSlot slot,
+            NpgsqlTestDecodingPluginOptions options = default, NpgsqlLogSequenceNumber? walLocation = null, CancellationToken cancellationToken = default)
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return StartReplicationInternal(connection, slot, options, walLocation, cancellationToken);
+        }
+
+        static async Task<IAsyncEnumerable<NpgsqlTestDecodingData>> StartReplicationInternal(
+            NpgsqlLogicalReplicationConnection connection, NpgsqlTestDecodingReplicationSlot slot,
+            NpgsqlTestDecodingPluginOptions options = default, NpgsqlLogSequenceNumber? walLocation = null, CancellationToken cancellationToken = default)
+        {
+            var stream = await connection.StartReplicationForPlugin(slot, walLocation, options.GetOptionPairs(), cancellationToken);
+#pragma warning disable CA2016
+            // ReSharper disable once MethodSupportsCancellation
+            return StreamData(stream, connection.Encoding!);
+#pragma warning restore CA2016
+        }
+
+        static async IAsyncEnumerable<NpgsqlTestDecodingData> StreamData(IAsyncEnumerable<NpgsqlXLogDataMessage> messageStream, Encoding encoding, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            await foreach (var msg in messageStream.WithCancellation(cancellationToken))
+                yield return new NpgsqlTestDecodingData(msg.WalStart, msg.WalEnd, msg.ServerClock, await GetString(msg.Data));
+
+            async Task<string> GetString(Stream xlogDataStream)
+            {
+                var memoryStream = new MemoryStream();
+                await xlogDataStream.CopyToAsync(memoryStream, 4096, CancellationToken.None);
```

Comment:
> does this mean we truncate the text if it's more than 4096?

No it's only the buffer size for copying. I'm using this overload because it's the only one that is available in all of out target platforms.

Reply:
Oh sorry, misread the code. I don't really care much about the efficiency of test_decoding in any case :)